scene/resources/paragraph_layout.cpp

```diff
@@ -0,0 +1,531 @@
+		return Point2(0, 0);
+
+	return glyph_offsets[p_index];
+}
+
+TextDirection ClusterInfo::get_direction() const {
+
+	return direction;
+}
+
+int ClusterInfo::get_min_char_index() const {
+
+	return min_index;
+}
+
+int ClusterInfo::get_max_char_index() const {
+
+	return max_index;
+}
+
+int ClusterInfo::get_offset() const {
+
+	return offset;
+}
+
+int ClusterInfo::get_length() const {
+
+	return length;
+}
+
+int ClusterInfo::get_leading_edge() const {
+	if (direction == DIR_LTR) {
+		return offset + length;
+	} else {
+		return offset;
+	}
+}
+
+int ClusterInfo::get_tailing_edge() const {
+	if (direction == DIR_LTR) {
+		return offset;
+	} else {
+		return offset + length;
+	}
+}
+
+int ClusterInfo::get_left_edge() const {
+	return offset;
+}
+
+int ClusterInfo::get_right_edge() const {
+	return offset + length;
+}
+
+/*************************************************************************/
+
+void ParagraphLayout::set_text(const String &p_string, TextDirection p_direction) {
+
+	UErrorCode err = U_ZERO_ERROR;
+
+	base_direction = (p_direction == DIR_MIXED) ? DIR_AUTO : p_direction;
+	utf16_text.clear();
+
+	if (p_string.length() > 0) {
+		if (sizeof(CharType) == 4) {
+			//:'‑( wchar_t why???
+			int32_t utf16_length;
+			u_strFromUTF32(NULL, 0, &utf16_length, (const UChar32 *)p_string.ptr(), p_string.length(), &err);
+			utf16_text.resize(utf16_length);
+			err = U_ZERO_ERROR;
+			u_strFromUTF32((UChar *)utf16_text.ptr(), utf16_length, &utf16_length, (const UChar32 *)p_string.ptr(), p_string.length(), &err);
+		} else {
+			utf16_text.resize(p_string.length());
+			memcpy((UChar *)utf16_text.ptr(), p_string.ptr(), p_string.length() * sizeof(UChar));
+		}
+	}
+	dirty = true;
+}
+
+void ParagraphLayout::set_max_width(int p_max_width) {
+
+	max_width = p_max_width;
+	dirty = true;
+}
+
+void ParagraphLayout::set_font(const Ref<Font> &p_font) {
+
+	font = p_font;
+	dirty = true;
+}
+
+void ParagraphLayout::set_ot_features(const String &p_features) {
+
+	features.clear();
+
+	Vector<String> v_features = p_features.split(",");
+	for (int i = 0; i < v_features.size(); i++) {
+		hb_feature_t feature;
+		if ((v_features[i].length() == 5) && (v_features[i].begins_with("-") || v_features[i].begins_with("!"))) {
+			if (hb_feature_from_string(v_features[i].substr(1, 4).ascii().get_data(), 4, &feature)) {
+				feature.value = 0;
+				features.push_back(feature);
+			}
+		} else if (v_features[i].length() == 4) {
+			if (hb_feature_from_string(v_features[i].substr(0, 4).ascii().get_data(), 4, &feature)) {
+				feature.value = 1;
+				features.push_back(feature);
+			}
+		} else if (v_features[i].length() > 5) {
+			Vector<String> feature_rng = v_features[i].split(":");
+			if (feature_rng.size() == 3) {
+				if (feature_rng[1].is_valid_integer() && feature_rng[2].is_valid_integer()) {
+					if ((feature_rng[0].length() == 5) && (feature_rng[0].begins_with("-") || feature_rng[0].begins_with("!"))) {
```

Comment:
`hb_feature_from_string()` should correctly handle feature negation and ranges already, see `hb-view --help-features` for the supported syntax.